### ISSUE TITLE
Consolidate GQA optimization: add packed QKV support and improve test coverage

### ISSUE THREAD

**Issue by titaiwangms** [onnxruntime/mobius#11]
# Consolidate GQA optimization: add packed QKV support and improve test coverage

## Before You Start

1. **Read the skill**: `.github/skills/writing-rewrite-rules/SKILL.md` — covers the `RewriteRuleClassBase` API, pattern matching, `check()`/`rewrite()` methods, file organization, and testing conventions.
2. **Read the existing rule**: `src/mobius/rewrite_rules/_group_query_attention.py` (224 lines) — this is the rule you will extend.
3. **Read its tests**: `src/mobius/rewrite_rules/_group_query_attention_test.py` (156 lines) — 7 existing tests.
4. **Read the ORT GQA spec**: https://github.com/microsoft/onnxruntime/blob/main/docs/ContribOperators.md — search for `GroupQueryAttention`. Note the packed QKV input format.

## Repo Conventions

```bash
# Install for development
pip install -e ".[transformers,testing]"

# Run rewrite rule tests only (~fast)
python -m pytest src/mobius/rewrite_rules/ -q --tb=short

# Run GQA tests specifically
python -m pytest src/mobius/rewrite_rules/_group_query_attention_test.py -sv

# Run full non-integration test suite
python -m pytest tests/build_graph_test.py tests/cli_test.py src/ -q \
  -k "not phi4mm and not apply_weights_unknown" --tb=short -n auto

# Lint and auto-format
lintrunner f --output oneline --all-files
```

- `from __future__ import annotations` in every module
- Ruff for linting/formatting (line-length=95, Python 3.10+)
- MyPy strict mode (excludes `*_test.py`)
- Git: linear commit history, always `--signoff`
- Test files must end in `_test.py`
- Co-located tests: `src/mobius/**/*_test.py` next to source

### IN scope
- Extend `RotaryAttentionToGQA` in `_group_query_attention.py` to attempt packed QKV, falling back to separate Q/K/V when packing is not possible
- Tests for packed QKV behavior (added to existing `_group_query_attention_test.py`)

### OUT of scope — do NOT implement
- Static cache + GQA (the GQA pattern requires `past_key`/`past_value` graph inputs and `attention_mask`; static cache uses `TensorScatter` + `nonpad_kv_seqlen` instead — the pattern cannot match)
- Creating new rule files or new rule classes — all changes go in the existing file and class
- Benchmark / arch diff entries for GQA-optimized models
- `optimize` parameter on `build_from_module()` API
- Changes to the Attention component or decoder layer
- New model architectures

### Existing Rule: `RotaryAttentionToGQA`

**File**: `src/mobius/rewrite_rules/_group_query_attention.py`
**Class**: `RotaryAttentionToGQA(RewriteRuleClassBase)` (line 31)

**What it matches** (`pattern()` at line 74):
```python
def pattern(self, op, q_pre, k_pre, v, attention_bias, past_key, past_value, cos, sin):
    q_rot = op.RotaryEmbedding(q_pre, cos, sin, _allow_other_attributes=True)
    k_rot = op.RotaryEmbedding(k_pre, cos, sin, _allow_other_attributes=True)
    return op.Attention(
        q_rot, k_rot, v, attention_bias, past_key, past_value,
        _allow_other_attributes=True,
        _outputs=["attn_out", "present_key", "present_value"],
    )
```

**What it checks** (`check()` at line 100):
- `scale`, `q_num_heads`, `kv_num_heads` attributes exist on the Attention node
- `cos`/`sin` are produced by `Gather` nodes (position-indexed cache tables)
- `past_key`/`past_value` are graph inputs (not computed — filters out vision encoder)

**What it produces** (`rewrite()` at line 132):
```python
outputs = op.op_multi_out(
    "GroupQueryAttention",
    inputs=[q_pre, k_pre, v, past_key, past_value,
            self._seqlens_k, self._total_seq_len,
            self._cos_cache, self._sin_cache],
    domain="com.microsoft",
    attributes={
        "num_heads": q_num_heads,
        "kv_num_heads": kv_num_heads,
        "scale": scale,
        "do_rotary": 1,
        "rotary_interleaved": 0,
    },
    num_outputs=3,
)
```

**Shared state** (computed once, reused across layers):
- `self._seqlens_k` = `Cast(ReduceSum(attention_mask, axis=1) - 1, INT32)` (line 167-174)
- `self._total_seq_len` = `Cast(Gather(Shape(attention_mask), 1), INT32)` (line 176-182)
- `self._cos_cache` / `self._sin_cache` = traced from cos/sin Gather inputs (line 154-156)

### Existing Tests: 7 passing

**File**: `src/mobius/rewrite_rules/_group_query_attention_test.py`

| Test | What it verifies |
|------|-----------------|
| `test_rules_returns_rule_set` | Returns `RewriteRuleSet` |
| `test_replaces_attention_with_gqa` | 28 Attention → 0, 28 GQA (Qwen3-0.6B) |
| `test_absorbs_rotary_embedding` | 56 RotaryEmbedding → 0 |
| `test_absorbs_rotary_without_qk_norm` | 30 Attention → 30 GQA (SmolLM2, no QK norm) |
| `test_preserves_non_matching_model` | Vision encoder Attention untouched |
| `test_rewritten_model_runs_with_ort` | Tiny model → GQA rewrite → ORT inference → correct output shape |
| `test_combined_gqa_and_skip_norm_runs_with_ort` | GQA + SkipNorm → valid ORT model |

### Graph Pattern (what the ONNX graph actually looks like per layer)

For a standard causal LM (e.g., llama, qwen2) built with `build_from_module()`:

```
hidden_states                           # [B, seq, hidden_size]
  ├─ Transpose → MatMul (q_proj)        # hidden → q_dim
  ├─ Transpose → MatMul (k_proj)        # hidden → kv_dim
  └─ Transpose → MatMul (v_proj)        # hidden → kv_dim
       │              │           │
  RotaryEmbedding  RotaryEmbedding  │
       │              │           │
       └──── Attention(q, k, v, attn_bias, past_k, past_v) ────→ out, present_k, present_v
                                  │
                     Transpose → MatMul (o_proj)
```

Note: The `Transpose` before each `MatMul` is because weights are stored transposed (fused_matmul rule can handle this separately). The `q_proj`, `k_proj`, `v_proj` are separate `Linear` modules producing separate `MatMul` nodes.

### Extend `RotaryAttentionToGQA` with Packed QKV

**Goal**: Modify the existing `RotaryAttentionToGQA` class in `src/mobius/rewrite_rules/_group_query_attention.py` so that its `rewrite()` method attempts to pack Q/K/V into a single fused QKV weight. If packing is not possible (QK norm present, bias complications, incompatible shapes), it falls back to the current separate Q/K/V behavior. Single rule, two paths — no new file, no new class.

**File to modify**: `src/mobius/rewrite_rules/_group_query_attention.py`

**Design — single rule, two code paths in `rewrite()`**:

The `pattern()` and `check()` methods stay the same — they already match the `RotaryEmbedding + Attention` pattern. The change is in `rewrite()`:

```python
def rewrite(self, op, q_pre, k_pre, v, attention_bias, past_key, past_value,
            cos, sin, attn_out, present_key, present_value, **_):
    # ... existing setup (get attributes, trace cos/sin, compute seqlens_k) ...

    # TRY packed QKV path
    packed_qkv = self._try_pack_qkv(op, q_pre, k_pre, v)
    if packed_qkv is not None:
        # Packed path: single QKV input, remaining Q/K/V slots are None
        outputs = op.op_multi_out(
            "GroupQueryAttention",
            inputs=[packed_qkv, None, None, past_key, past_value,
                    self._seqlens_k, self._total_seq_len,
                    self._cos_cache, self._sin_cache],
            domain="com.microsoft",
            attributes={...same attributes...},
            num_outputs=3,
        )
    else:
        # Fallback: separate Q, K, V (current behavior — unchanged)
        outputs = op.op_multi_out(
            "GroupQueryAttention",
            inputs=[q_pre, k_pre, v, past_key, past_value,
                    self._seqlens_k, self._total_seq_len,
                    self._cos_cache, self._sin_cache],
            domain="com.microsoft",
            attributes={...same attributes...},
            num_outputs=3,
        )
    return outputs[0], outputs[1], outputs[2]
```

**What `_try_pack_qkv()` must do** (new private method on the class):

1. **Trace back** from `q_pre`, `k_pre`, `v` to their producing MatMul ops
2. **Verify all 3 MatMuls** share the same first input (`hidden_states`)
3. **Verify no QK norm** — check that `q_pre` and `k_pre` come directly from MatMul, not through RMSNormalization/LayerNorm nodes
4. **Extract weight initializers** from all 3 MatMuls — they must be constant (`ir.Tensor`) values, not computed
5. **Handle Transpose** — weights may be transposed (Transpose → MatMul pattern from `Linear` module). Account for this when concatenating
6. **Concatenate weights**: `W_qkv = numpy.concatenate([W_q, W_k, W_v], axis=0)` (along output dim, accounting for transpose)
7. **Handle bias** if present: concatenate biases similarly, or return `None` to fall back
8. **Create packed MatMul**: single `MatMul(hidden, W_qkv)` (or `FusedMatMul` with `transB=1` if weights were transposed)
9. **Return** the packed QKV tensor, or `None` if packing failed at any step

**When packing should fail (return None)**:
- `q_pre`, `k_pre`, or `v` don't come from MatMul nodes
- The 3 MatMuls don't share the same input
- QK norm nodes exist between MatMul and RotaryEmbedding (e.g., Qwen3 with `attn_qk_norm=True`)
- Weight initializers are not available as constants
- Shape incompatibility

**Success criteria**:
```python
# After applying group_query_attention_rules() to a llama/qwen2 model (no QK norm):
assert count_ops(model).get("Attention", 0) == 0
assert count_ops(model).get("RotaryEmbedding", 0) == 0
assert count_ops(model)["GroupQueryAttention"] == num_layers
# MatMul count should decrease by 2 * num_layers (3 separate → 1 packed per layer)
assert count_ops(model)["MatMul"] == original_matmul_count - 2 * num_layers

# After applying group_query_attention_rules() to a qwen3 model (HAS QK norm):
# Packing fails gracefully → falls back to separate Q/K/V GQA (current behavior)
assert count_ops(model).get("Attention", 0) == 0
assert count_ops(model)["GroupQueryAttention"] == num_layers
# MatMul count stays the same (no packing occurred)
assert count_ops(model)["MatMul"] == original_matmul_count

# ALL 7 existing tests must continue to pass unchanged
# Model must still run with ORT and produce correct output shape
session = OnnxModelSession(model)
result = session.run(feeds)
assert result["logits"].shape == (batch, seq, vocab)
```

**Gotchas**:
- The `FusedMatMul` rule may have already replaced `Transpose+MatMul` with `FusedMatMul(transB=1)`. Your `_try_pack_qkv()` must handle BOTH cases: trace through `Transpose+MatMul` AND `FusedMatMul`. Or: document that GQA rule should be applied BEFORE fused_matmul (simpler approach).
- Weight concatenation must happen in the `rewrite()` method. Use `numpy` to concatenate the constant values. Create a new `ir.Tensor` for the packed weight.
- ORT's GQA expects packed QKV in layout `[q_dim + kv_dim + kv_dim]` (Q first, then K, then V). Verify this against the ORT spec.
- All 7 existing tests must still pass. The packed path is additive — it kicks in when conditions allow, falls back otherwise.
- No new rule class, no new rule file, no new registration in `__init__.py` or CLI `rule_map`. The rule name stays `group_query_attention`.

### Tests for Packed QKV (add to existing test file)

**File to modify**: `src/mobius/rewrite_rules/_group_query_attention_test.py`

**Add these test cases** to the existing test class:

```python
@pytest.mark.parametrize("model_type", ["llama", "qwen2", "phi3"])
def test_packed_qkv_on_all_compatible_models(self, model_type):
    """Build each model with tiny config, apply rule, verify packing succeeded:
    - 0 Attention ops remain
    - 0 RotaryEmbedding ops remain
    - GQA ops == num_layers
    - MatMul count decreased by 2 * num_layers (3 separate → 1 packed)
    Catches model-specific edge cases across architectures.
    """

def test_packed_weight_shape_is_correct(self):
    """Build tiny llama, apply rule, inspect the packed W_qkv initializer.
    - Verify shape is [hidden, q_dim + 2*kv_dim] (or transposed equivalent)
    - Verify the original 3 separate q_proj/k_proj/v_proj MatMul nodes are gone
    - Verify the packed weight contains the correct concatenation of Q, K, V weights
    """

def test_falls_back_to_separate_qkv_with_qk_norm(self):
    """Build qwen3 (attn_qk_norm=True), apply rule.
    QK norm prevents packing — rule falls back to separate Q/K/V.
    GQA ops should still be produced, but MatMul count unchanged.
    """

def test_packed_model_runs_with_ort(self):
    """Build tiny llama model, apply rule, fill weights, run with ORT.
    Verify output shape matches expected (batch, seq, vocab).
    """

def test_combined_packed_gqa_and_skip_norm(self):
    """Apply GQA (with packing) + skip_norm, verify ORT inference works."""
```

**Use existing test helpers** from `src/mobius/rewrite_rules/_testing_utils.py`:
- `count_ops(model)` — returns `dict[str, int]` of op type counts
- `fill_random_weights(model)` — fills all initializers with random data
- `make_prefill_feeds(session)` — creates prefill input feeds for ORT

## Key Files Map

| File | Lines | What to do |
|------|-------|-----------|
| `src/mobius/rewrite_rules/_group_query_attention.py` | 224 | **MODIFY** — add `_try_pack_qkv()` method, update `rewrite()` to try-pack-first |
| `src/mobius/rewrite_rules/_group_query_attention_test.py` | 156 | **MODIFY** — add 6 new test cases for packed QKV behavior (13 total) |
| `src/mobius/rewrite_rules/_testing_utils.py` | ~60 | READ ONLY — use `count_ops`, `fill_random_weights`, `make_prefill_feeds` |
| `.github/skills/writing-rewrite-rules/SKILL.md` | — | READ FIRST — rewrite rule conventions |

### Build Pipeline (for reference)

```
build("model_id")
  → HF config → registry lookup → module instantiation
  → build_from_module(module, config, task)
    → task.build() → graph construction (forward() builds ONNX nodes)
    → _optimize() → 5 default IR passes (2 iterations):
        IdentityEliminationPass, LiftConstantsToInitializersPass,
        CommonSubexpressionEliminationPass, RemoveUnusedNodesPass,
        RemoveUnusedOpsetsPass
  → set graph names → download + apply weights → return ModelPackage
```

### Rewrite Rule API (onnxscript.rewriter)

```python
class MyRule(RewriteRuleClassBase):
    def pattern(self, op, ...):
        """Return the output node(s) of the pattern to match."""
        return op.SomeOp(...)

    def check(self, context, **bindings) -> MatchResult:
        """Return MatchResult — call .fail('reason') to reject a match."""
        result = MatchResult()
        if bad_condition:
            return result.fail("why")
        return result

    def rewrite(self, op, **bindings):
        """Return replacement node(s). Same output count as pattern."""
        return op.op_multi_out("NewOp", inputs=[...], domain="com.microsoft", ...)

def my_rules() -> RewriteRuleSet:
    return RewriteRuleSet([MyRule().rule()])

# Usage:
from onnxscript.rewriter import rewrite
rewrite(model, pattern_rewrite_rules=my_rules())
```

### Static Cache — Why It's Out of Scope

Static cache (PR #9) produces this Attention pattern:
```
TensorScatter(key_cache, k_new, write_indices)
TensorScatter(value_cache, v_new, write_indices)
Attention(q, updated_k, updated_v, None, None, None, nonpad_kv_seqlen,
          is_causal=1, ...)
```

The GQA rewrite checks `past_key.producer() is not None` → fails (past_key is a TensorScatter output, not a graph input). Also, no `attention_mask` graph input exists to compute `seqlens_k`. Completely different graph topology — needs a separate rule if/when ORT GQA supports it.

## References

- [ORT GroupQueryAttention spec](https://github.com/microsoft/onnxruntime/blob/main/docs/ContribOperators.md) — search for `GroupQueryAttention`
- PR #9: Static cache support (StaticCacheCausalLMTask, StaticCacheState) — merged
- `.github/skills/writing-rewrite-rules/SKILL.md` — read before starting
- `onnxscript.rewriter` source: used for `rewrite()`, `RewriteRuleClassBase`, `RewriteRuleSet`, `MatchResult`
